src/Sema.zig

```diff
@@ -22923,7 +22943,7 @@ fn beginComptimePtrLoad(
                 if (coerce_in_mem_ok) {
                     const payload_val = switch (ptr_val.tag()) {
                         .eu_payload_ptr => tv.val.castTag(.eu_payload).?.data,
-                        .opt_payload_ptr => tv.val.castTag(.opt_payload).?.data,
+                        .opt_payload_ptr => if (tv.val.castTag(.opt_payload)) |some| some.data else tv.val,
```

Comment:
I think we also need to a null check here, for the following case:
```zig
comptime {
    var val: u8 = 15;
    var opt_ptr: ?*const u8 = &val;

    const payload_ptr = &opt_ptr.?;
    opt_ptr = null;
    _ = payload_ptr.*;
}
```

Reply:
In that test case assigning `null` to `opt_ptr` does not affect `payload_ptr` in any way since it points to the still valid `val`.

Reply:
To see what I mean, add another `.*` to payload_ptr on the last line

It crashes the compiler, because `payload_ptr.*` yields `null`

Reply:
Hmm, I'm not sure what the expected behavior is in that case. Can't compare to stage1 either since it also just crashes.

Reply:
I'll add an error using the same messages as the safety checks on `.?` and `catch unreachable`.

Reply:
I think this is out of scope for this pr since it is pretty obvious that this case should fail since `?*const u8` and `*const u8` have the same representation both at runtime and at comptime:
```zig
comptime {
    var val: u8 = 15;
    var opt_ptr: ?*const u8 = &val;

    const payload_ptr = &opt_ptr.?;
    opt_ptr = null;
    _ = payload_ptr.*.*;
}
```

But what about these two cases? At runtime they are represented as separate fields in a struct and changing the status does not alter the payload while at comptime they take up the same space.
```zig
comptime {
    var opt: ?u8 = 15;

    const payload_ptr = &opt.?;
    opt = null;
    @compileLog(payload_ptr.*); // @as(u8, null)??
}
comptime {
    var val: u8 = 15;
    var err_union: anyerror!u8 = val;

    const payload_ptr = &(err_union catch unreachable);
    err_union = error.Foo;
    _ = payload_ptr.*; // error: attempt to unwrap error: Foo
}
```

cc @SpexGuy.

Reply:
I made it an error for now since it prevents a crash but the desired behavior should be clarified somewhere.

For reference, at runtime the first example segfaults and the latter two result in zero.